amy/workshops/models.py

```diff
@@ -2671,6 +2671,34 @@ def clean_event(
         elif not requirement.event_required and self.event:
             return self.get_not_required_error(requirement)
 
+        if self.event and hasattr(self, "trainee") and self.trainee:
+            # check this trainee has a learner task for this event
+            try:
+                Task.objects.get(
+                    person=self.trainee,
+                    event=self.event,
+                    role=Role.objects.get(name="learner"),
+                )
```

Comment:
One concern I have about this code is that we're using a SELECT query during creation of training progress. Maybe such code could live in a form `clean_event` method?